nautobot_ssot/contrib.py

```diff
@@ -44,13 +45,54 @@ class ProviderModel(NautobotModel):
     name: str
 
 
+# This type describes a set of parameters to use as a dictionary key for the cache. As such, its needs to be hashable
+# and therefore a frozenset rather than a normal set or a list.
+#
+# The following is an example of a parameter set that describes a tenant based on its name and group:
+# frozenset(
+#  [
+#   ("name", "ABC Inc."),
+#   ("group__name", "Customers"),
+#  ]
+# )
+ParameterSet = FrozenSet[Tuple[str, Hashable]]
+
+
 class NautobotAdapter(DiffSync):
     """
     Adapter for loading data from Nautobot through the ORM.
 
     This adapter is able to infer how to load data from Nautobot based on how the models attached to it are defined.
     """
 
+    # This dictionary acts as an opt-in ORM cache.
+    _cache: DefaultDict[str, Dict[ParameterSet, Model]]
+    _cache_hits: DefaultDict[str, int] = defaultdict(int)
+
+    def __init__(self, *args, **kwargs):
+        """Invalidate cache in __init__."""
+        super().__init__(*args, **kwargs)
+        self.invalidate_cache()
+
+    def invalidate_cache(self, zero_out_hits=True):
+        """Invalidates all the objects in the ORM cache."""
+        self._cache = defaultdict(dict)
+        if zero_out_hits:
+            self._cache_hits = defaultdict(int)
+
+    def get_from_orm_cache(self, parameters: Dict, model_class: Type[Model]):
```

Comment:
This is good and covers many use cases, but there are situations where I do not want my cache lookup key to be tied to how the object(s) are queried. Perhaps I have some complex query logic and I just want a simple key/value cache that is disconnected from my actual query.